### DISCUSSION TITLE
Getting error "AttributeError: module 'taichi' has no attribute 'normalized"

### DISCUSSION THREAD

**Discussion by jchinniah** [community/community#107475] [Programming Help]
### Select Topic Area

Question

### Body

Used the following code from Ye Kuang
import taichi as ti
import numpy as np
from taichi.math import *
sz = 800
res = (sz, sz)

ti.init(arch=ti.gpu)  # starts cuda
color_buffer = ti.Vector.field(3, dtype=ti.f32, shape=res)
aspect_ratio = res[0] / res[1]

eps = 1e-4
inf = 1e10
fov = 0.8
max_ray_depth = 10

mat_none = 0
mat_lambertian = 1
mat_light = 2

camera_pos = ti.Vector([0.0, 0.6, 3.0])
light_color = ti.Vector([0.9, 0.85, 0.7])

ti.func
def ray_plane_intersect(pos, d, pt_on_plane, norm):
    t = inf
    denom = ti.dot(d, norm)
    if abs(denom) > eps:
        t = ti.dot((pt_on_plane - pos), norm) / denom
    return t

@ti.func
def intersect_scene(ray_o, ray_d):
    closest, normal = inf, ti.Vector.zero(ti.f32, 3)
    color, mat = ti.Vector.zero(ti.f32, 3), mat_none

    # left
    pnorm = ti.Vector([1.0, 0.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([-1.1, 0.0, 0.0]),
                                   pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = ti.Vector([0.65, 0.05, 0.05]), mat_lambertian
    # right
    pnorm = ti.Vector([-1.0, 0.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([1.1, 0.0, 0.0]),
                                   pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = ti.Vector([0.12, 0.45, 0.15]), mat_lambertian
    # bottom
    gray = ti.Vector([0.93, 0.93, 0.93])
    pnorm = ti.Vector([0.0, 1.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 0.0, 0.0]),
                                   pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = gray, mat_lambertian
    # top
    pnorm = ti.Vector([0.0, -1.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 2.0, 0.0]),
                                   pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = gray, mat_lambertian
    # far
    pnorm = ti.Vector([0.0, 0.0, 1.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 0.0, 0.0]),
                                   pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = gray, mat_lambertian

    return closest, normal, color, mat

ti.func
def random_in_unit_sphere():
    p = ti.Vector([0.0, 0.0, 0.0])
    while True:
        for i in ti.static(range(3)):
            p[i] = ti.random() * 2 - 1.0
        if p.dot(p) <= 1.0:
            break
    return p

@ti.func
def sample_ray_dir(indir, normal, hit_pos, mat):
    return ti.normalized(random_in_unit_sphere() + normal)

@ti.kernel
def render():
    for u, v in color_buffer:
        ray_pos = camera_pos
        ray_dir = ti.Vector([
            (2 * fov * (u + ti.random()) / res[1] - fov * aspect_ratio),
            (2 * fov * (v + ti.random()) / res[1] - fov),
            -1.0,
        ])
        ray_dir = ti.normalized(ray_dir)

        px_color = ti.Vector([0.0, 0.0, 0.0])
        throughput = ti.Vector([1.0, 1.0, 1.0])
        depth = 0

        while depth < max_ray_depth:
            closest, hit_normal, hit_color, mat = intersect_scene(ray_pos, ray_dir)
            if mat == mat_none:
                break
            if mat == mat_light:
                px_color = throughput * light_color
                break
            hit_pos = ray_pos + closest * ray_dir
            depth += 1
            ray_dir = sample_ray_dir(ray_dir, hit_normal, hit_pos, mat)
            ray_pos = hit_pos + eps * ray_dir
            throughput *= hit_color

        color_buffer[u, v] += px_color
 
def main():

   gui = ti.GUI('Cornell Box', res)
   for i in range(50000):
        render()
        img = color_buffer.to_numpy()
        gui.set_image(img)
        gui.show()

if __name__ == "__main__":
    main()

**Comment by ZephyrusBlaze** [comment]
Here is the corrected code:
```py
import taichi as ti
import numpy as np
# taichi.math has not been used

sz = 800
res = (sz, sz)

ti.init(arch=ti.gpu)  # Start CUDA
color_buffer = ti.Vector.field(3, dtype=ti.f32, shape=res)
aspect_ratio = res[0] / res[1]

eps = 1e-4
inf = 1e10
fov = 0.8
max_ray_depth = 10

mat_none = 0
mat_lambertian = 1
mat_light = 2

camera_pos = ti.Vector([0.0, 0.6, 3.0])
light_color = ti.Vector([0.9, 0.85, 0.7])


@ti.func
def ray_plane_intersect(pos, d, pt_on_plane, norm):
    t = inf
    denom = d.dot(norm)
    if abs(denom) > eps:
        t = (pt_on_plane - pos).dot(norm) / denom
    return t


@ti.func
def intersect_scene(ray_o, ray_d):
    closest, normal = inf, ti.Vector([0.0, 0.0, 0.0])
    color, mat = ti.Vector([0.0, 0.0, 0.0]), mat_none

    # Left
    pnorm = ti.Vector([1.0, 0.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([-1.1, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = ti.Vector([0.65, 0.05, 0.05]), mat_lambertian
    # Right
    pnorm = ti.Vector([-1.0, 0.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([1.1, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = ti.Vector([0.12, 0.45, 0.15]), mat_lambertian
    # Bottom
    gray = ti.Vector([0.93, 0.93, 0.93])
    pnorm = ti.Vector([0.0, 1.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = gray, mat_lambertian
    # Top
    pnorm = ti.Vector([0.0, -1.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 2.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = gray, mat_lambertian
    # Far
    pnorm = ti.Vector([0.0, 0.0, 1.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = gray, mat_lambertian

    return closest, normal, color, mat


@ti.func
def random_in_unit_sphere():
    p = ti.Vector([0.0, 0.0, 0.0])
    while True:
        for i in ti.static(range(3)):
            p[i] = ti.random() * 2 - 1.0
        if p.norm_sqr() <= 1.0:
            break
    return p


@ti.func
def sample_ray_dir(indir, normal, hit_pos, mat):
    return (random_in_unit_sphere() + normal).normalized()


@ti.kernel
def render():
    for u, v in color_buffer:
        ray_pos = camera_pos
        ray_dir = ti.Vector([
            (2 * fov * (u + ti.random()) / res[1] - fov * aspect_ratio),
            (2 * fov * (v + ti.random()) / res[1] - fov),
            -1.0,
        ])
        ray_dir = ray_dir.normalized()

        px_color = ti.Vector([0.0, 0.0, 0.0])
        throughput = ti.Vector([1.0, 1.0, 1.0])
        depth = 0

        while depth < max_ray_depth:
            closest, hit_normal, hit_color, mat = intersect_scene(ray_pos, ray_dir)
            if mat == mat_none:
                break
            if mat == mat_light:
                px_color = throughput * light_color
                break
            hit_pos = ray_pos + closest * ray_dir
            depth += 1
            ray_dir = sample_ray_dir(ray_dir, hit_normal, hit_pos, mat)
            ray_pos = hit_pos + eps * ray_dir
            throughput *= hit_color

        color_buffer[u, v] += px_color


def main():
    gui = ti.GUI('Cornell Box', res)
    for i in range(50000):
        render()
    img = color_buffer.to_numpy()
    gui.set_image(img)
    gui.show()


if __name__ == "__main__":
    main()
```

Fixed attribute errors by using proper method calls (`dot()`, `norm_sqr()`, `normalized()`) instead of `taichi.dot()`, `taichi.norm_sqr()`, `taichi.normalized()`.

**Reply by jchinniah** [reply]
Thanks Om Sahu (TheStrange-007) for your quick help. No attribute error now, but I get a white screen. I have to figure out why Cornell box is not visible. Thanks a lot again.

**Reply by jchinniah** [reply]
I am getting the Cornell box now. But it has some black pixels in the wall. Image location at the bottom of the code

Here is the updated code.

import taichi as ti
import numpy as np
# taichi.math has not been used
#Fixed attribute errors by using proper method calls (dot(), norm_sqr(), normalized()) 
#instead of taichi.dot(), taichi.norm_sqr(), taichi.normalized().#

sz = 500
res = (sz, sz)

ti.init(arch=ti.gpu)  # Start CUDA
color_buffer = ti.Vector.field(3, dtype=ti.f32, shape=res)
aspect_ratio = res[0] / res[1]

eps = 1e-4
inf = 1e10
fov = 0.8
max_ray_depth = 10

mat_none = 0
mat_lambertian = 1
mat_light = 2

camera_pos = ti.Vector([0.0, 0.6, 3.0])
light_color = ti.Vector([0.9, 0.85, 0.7])


@ti.func
def ray_plane_intersect(pos, d, pt_on_plane, norm):
    t = inf
    denom = d.dot(norm)
    if abs(denom) > eps:
        t = (pt_on_plane - pos).dot(norm) / denom
    return t


@ti.func
def intersect_scene(ray_o, ray_d):
    closest, normal = inf, ti.Vector([0.0, 0.0, 0.0])
    color, mat = ti.Vector([0.0, 0.0, 0.0]), mat_none

    # Left
    pnorm = ti.Vector([1.0, 0.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([-1.1, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = ti.Vector([0.65, 0.05, 0.05]), mat_lambertian

    # Right
    pnorm = ti.Vector([-1.0, 0.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([1.1, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = ti.Vector([0.12, 0.45, 0.15]), mat_lambertian

    # Bottom
    gray = ti.Vector([0.93, 0.93, 0.93])
    pnorm = ti.Vector([0.0, 1.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        color, mat = gray, mat_lambertian
        #color, mat = gray, mat_light  # mat_lambertian

    # Top
    pnorm = ti.Vector([0.0, -1.0, 0.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 2.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        #color, mat = gray, mat_lambertian
        color, mat = gray, mat_light  # mat_lambertian

    # Far
    pnorm = ti.Vector([0.0, 0.0, 1.0])
    cur_dist = ray_plane_intersect(ray_o, ray_d, ti.Vector([0.0, 0.0, 0.0]), pnorm)
    if 0 < cur_dist < closest:
        closest = cur_dist
        normal = pnorm
        #color, mat = gray, mat_lambertian
        color, mat = gray, mat_light  # mat_lambertian

        #print(color)
       
    return closest, normal, color, mat


@ti.func
def random_in_unit_sphere():
    p = ti.Vector([0.0, 0.0, 0.0])
    while True:
        for i in ti.static(range(3)):
            p[i] = ti.random() * 2 - 1.0
        if p.norm_sqr() <= 1.0:
            break
    return p


@ti.func
def sample_ray_dir(indir, normal, hit_pos, mat):
    return (random_in_unit_sphere() + normal).normalized()


@ti.kernel
def render():
    for u, v in color_buffer:
        ray_pos = camera_pos
        ray_dir = ti.Vector([
            (2 * fov * (u + ti.random()) / res[1] - fov * aspect_ratio),
            (2 * fov * (v + ti.random()) / res[1] - fov),
            -1.0,
        ])
        ray_dir = ray_dir.normalized()

        px_color = ti.Vector([0.0, 0.0, 0.0])
        throughput = ti.Vector([1.0, 1.0, 1.0])
        depth = 0

        while depth < max_ray_depth:
            closest, hit_normal, hit_color, mat = intersect_scene(ray_pos, ray_dir)
            #print(hit_color)
            if mat == mat_none:
                break
            if mat == mat_light:
                px_color = throughput * light_color
                break
            hit_pos = ray_pos + closest * ray_dir
            depth += 1
            
            ray_dir = sample_ray_dir(ray_dir, hit_normal, hit_pos, mat)
            ray_pos = hit_pos + eps * ray_dir
            throughput *= hit_color
            
        color_buffer[u,v] += px_color       
        color_buffer[u,v] = color_buffer[u,v].normalized()

gui = ti.GUI('Cornell Box', res)
for i in range(5000):
    render()
    img = color_buffer.to_numpy()
    gui.set_image(img)
    gui.show()    


 
![image](https://github.com/community/community/assets/47703942/7b741dfa-c99f-4d8f-bb56-4309695eb000)

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @jchinniah ! We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Discussions category is solely related to conversations around the GitHub product Discussions. 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Reply by jchinniah** [reply]
Ok thanks
